packages/expo-updates/android/src/main/java/expo/modules/updates/selectionpolicy/SelectionPolicyUtils.java

```diff
@@ -0,0 +1,45 @@
+package expo.modules.updates.selectionpolicy;
+
+import android.util.Log;
+
+import org.json.JSONObject;
+
+import java.util.Iterator;
+
+import expo.modules.updates.db.entity.UpdateEntity;
+
+public class SelectionPolicyUtils {
+
+  public static final String TAG = SelectionPolicyUtils.class.getSimpleName();
+
+  public static boolean matchesFilters(UpdateEntity update, JSONObject manifestFilters) {
+    if (manifestFilters == null || update.metadata == null || !update.metadata.has("updateMetadata")) {
+      // empty matches all
+      return true;
+    }
+    try {
+      JSONObject updateMetadata = update.metadata.getJSONObject("updateMetadata");
+
+      // create lowercase copy for case-insensitive search
+      JSONObject metadataLCKeys = new JSONObject();
+      Iterator<String> metadataKeySet = updateMetadata.keys();
+      while (metadataKeySet.hasNext()) {
+        String key = metadataKeySet.next();
+        metadataLCKeys.put(key.toLowerCase(), updateMetadata.get(key));
+      }
+
+      Iterator<String> filterKeySet = manifestFilters.keys();
+      while (filterKeySet.hasNext()) {
+        String key = filterKeySet.next();
+        // once an update fails one filter, break early; we don't need to check the rest
+        if (metadataLCKeys.has(key) && !manifestFilters.get(key).equals(metadataLCKeys.get(key))) {
+          return false;
+        }
+      }
+    } catch (Exception e) {
+      Log.e(TAG, "Error filtering manifest using server data", e);
+      return true;
```

Comment:
Curious why this isn't `false`

If there is something unexpected wrong with an updates `updateMetadata` I would lean towards being suspect of the update in general and prefer the current (last working) update.

Reply:
That makes sense, but in this specific case I don't think it's possible for any unexpected JSONObject structure to actually cause an error to be thrown; we already know the object is valid JSON, we've checked before the `try` block that the `updateMetadata` key exists, and all the other operations are getting keys we know exist because we got them from the `keySet`. So in theory, we should never reach this catch block, and if we do, something really unexpected has gone wrong on our end -- rather than something being wrong with the update. So that's why I leaned towards `true` initially.

Happy to discuss/debate this further; I want to leave this as is, though, because I'd rather not change behavior in this PR (which is just refactoring).

Reply:
sounds good!